Initial git repo setup tasks

joakin opened the pull request:
This PR contains:

* [T172429: Initialize git repo](https://phabricator.wikimedia.org/T172429)
  * [x] .gitignore for node projects and dist/ folder
  * [x] package.json with name, description, etc
  * [x] Initial README.md version
  * [x] Docs folder with a changelog
* [T172430: Setup server entry point](https://phabricator.wikimedia.org/T172430)
  * [x] src/server/index.js
  * [x] Setup npm scripts in package.json
  * [x] Update README documenting the npm scripts
* [T172431: Setup client entrypoint](https://phabricator.wikimedia.org/T172431)
  * [x] src/client/index.js
  * [x] Output file name with hash (chunkhash)
  * [x] Change server/index to serve static assets from dist/ (express.static)
  * [x] Add ~~src/server/template/layout.ejs~~ src/server/template/page.js that generates the base HTML
  * [x] Server restarts after changes to .js templates on npm start
* [T172434: Setup linting and code formatting](https://phabricator.wikimedia.org/T172434)
  * [x] scripts lint & format
  * [x] Add .eslintrc.yml
  * [x] Extend wikimedia/eslint-config-node-services
  * [x] Extend eslint-config-prettier last to disable aesthetic rules
* [T172435: Setup git hooks](https://phabricator.wikimedia.org/T172435)
  *  [x] Setup a husky pre-commit hook that runs lint & test

joakin committed:
Add initial .gitignore

With linux/osx/node presets from gitignore.io

joakin committed:
Add initial package.json

joakin committed:
Doc: Add CHANGELOG in docs/ folder

joakin committed:
Add initial server entry point

...using express for later compatibility with the service-runner

Adds a / route that responds with Hello world and a fallback handler for
the rest of urls that sends back 404.

joakin committed:
Add npm start script

Which starts the server in development mode with file watching, so that
if you touch JS files it will restart the process automatically.

Update docs on README about running the project in development.

joakin committed:
Initial client side entry point

* Update .gitignore with dist/
* Add webpack for bundling
  * Add webpack.config.js for webpack configuration
    * Add assets-webpack-plugin for generating an asset manifest in
      dist/client/assets-manifest.json
    * Add clean-webpack-plugin to clean up the build folder
    * Use cheap source maps in dev, normal source-map in prod
* Add initial browser entry point in src/client/index.js

joakin committed:
Server: Serve static assets in dist/client/

joakin committed:
Add page template and serve it on /

Add a JS template with a HTML page layout that is served on "/".

The page HTML includes the client/index.js compiled file by reading the
assets-manifest.json and getting the file name with the hash.

joakin committed:
Run bundling on watch mode on npm start

For running two tasks in parallel across operating systems, we're using
npm-run-all. On npm start, we run all the `dev:*` tasks in parallel.

Additional changes:
* Add the production build step as npm run prod:build
* Specify dist/client and src/server/ as watched folders for the server
  to restart

joakin committed:
Add npm script "format" for auto formatting code

...using prettier with default settings.

joakin committed:
Add lint script to lint JS sources with eslint

Add eslint, the node-services eslint config, the prettier config, and
peer dependencies.

Configure eslint to allow parsing modern JS by setting the parser
options.

Linting errors in the JS sources are addressed separately next.

joakin committed:
Fix lint errors

joakin committed:
Run linting on pre-commit

Using husky to automatically install git hooks, run a precommit hoook to
lint the files in staging and avoid commiting unlinted files or failed
tests in the future.

In the next commit I'll add the formatting rules as linting errors so
that not-formatted files get caught by eslint.

joakin committed:
Add the formatter rules as eslint errors

In order to avoid not-formatted files being committed.

niedzielski commented on package.json:
```diff
@@ -0,0 +1,57 @@
+    "url": "https://phabricator.wikimedia.org/project/profile/2878/"
+  },
+  "homepage": "https://phabricator.wikimedia.org/project/profile/2878/",
+  "dependencies": {
```
I know this is controversial but I'd like to talk about fixing dependencies to exact patch versions. Semantic versioning only works when mistakes aren't made which is impossible.

joakin replied:
Let's talk about this later but we  are going to be using npm@5, which now has package-lock.json which actually fixes all the installed deps to the exact version. Like shrinkwrap did in previous npm versions. As you can see in the files changed, I've committed the package-lock.json to the repo, as recommended by npm inc., so that everyone npm installing will get **exactly** the same versions of the dependencies.

We will only upgrade by explicitly running `npm update` and then commiting the modified package-lock.json.

Very cool new feature from npm@5!! 🎉

niedzielski commented on .gitignore:
```diff
@@ -0,0 +1,110 @@
+
+dist/
+
+# Created by https://www.gitignore.io/api/osx,node,linux
```
This might be controversial but I would prefer we build this file ourselves. Here's my reasoning:

There's a lot of crufty boilterplate in this file we don't need. It's true that we may have to make several revisions to ensure everything is just right but I'd rather feel confident that this .gitignore file was accurate than just a listing of all the usual suspects. An accurate .gitignore reads easy, has less code, and can be very informative to other tools that also wish to identify unversioned files.

For example, consider the case of .DS_Store files:

> Patterns which a user wants Git to ignore in all situations (e.g., backup or temporary files generated by the user’s editor of choice) generally go into a file specified by core.excludesFile in the user’s ~/.gitconfig.

I think it would be better practice to place the exclude in a global Git [excludesfile](https://git-scm.com/docs/gitignore/2.9.1) per user. I actually don't mind us versioning project files for editors that are in use (for example, if you have WebStorm project file) but is there a case where any repo would ever _not_ want to ignore a .DS_Store file? For instance, I could see this file growing to include Windows' Thumbs.db and it kind of feels out of scope for our project.

As an example, albeit trivial, we've found that only three lines were necessary for the page library:
```git
/node_modules/                                                                  
/.eslintcache
/build/
```

And we've used this slim, precise .gitignore to help inform our bundler, watch, and ESLint ignore configs which files they shouldn't consider. And if we ever modify any of these files we'll reevaluate the others.

I got chewed out by a previous manager for expressing a similar sentiment in the past and he asked me if I thought I knew better than a community of contributors but... Come on, this just has tons of stuff in it that isn't relevant. gitignore.io and https://github.com/github/gitignore are good references but I suspect they're largely generated through a gradual amalgamation of StackOverflow answers without much careful planning or vision and should be considered for every repo but not necessarily copied verbatim. There are cases where file copies make sense (for example, artifacts that should be delivered with a library but aren't, tweaks to good code that has no hooks, etc) but I feel it's reasonable that we should know our own repo sufficiently well to identify what changes to ignore.

joakin replied:
Sounds good to me.

Addressed in 1a0b926

niedzielski commented on readme.md:
```diff
@@ -1,4 +1,13 @@
 
-[wiki]: https://www.mediawiki.org/wiki/Reading/Web/Projects/NewMobileWebsite
\ No newline at end of file
+## Development
```
This is awesome! Please also add an entry for our changelog like `##
 [Changelog](doc/changelog.md)` and possibly the license too `## [License](doc/license.md)`.

joakin replied:
Added link to the changelog in f8a8da4, lets do license when we include the license file in the repo.

I've created https://phabricator.wikimedia.org/T172682 to not forget about the license.

niedzielski commented on readme.md:
```diff
@@ -1,4 +1,13 @@
+That will run the server in development mode with file watching which will auto
+restart the server when files change.
+
+[wiki]: https://www.mediawiki.org/wiki/Reading/Web/Projects/NewMobileWebsite
```
What do you think about keeping the footnotes in the section they're referenced?

joakin replied:
Never done it, no preference here. Done in 2024588

niedzielski commented on package.json:
```diff
@@ -0,0 +1,57 @@
+    "dev:server": "NODE_ENV=development nodemon -w dist/client -w src/server/ src/server/index.js",
+    "prod:build": "NODE_ENV=production webpack -p",
+    "format": "prettier --write 'src/**/*.js'",
+    "lint": "eslint 'src/**/*.js'",
```
Neat! I didn't know that lint itself understood globs. I like that you've isolated this to its own script but consider the following configuration:

- `lint` script changes to just `eslint`.
- `lint:all` script is added which is `npm run -s lint 'src/**/*.js`.

This allows clients the flexibility to run with the default options plus their own custom flags like `npm run -s lint -- --fix FancyPants.js`.

Consider also (possibly in a follow-on PR that I am happy to write):
- Specify `--cache` on the ESLint command and adding `/.eslintcache` to the .gitignore.
- Specify also `--max-warnings 0` on the ESLint command. Warnings are useful during development for style-only concerns like whitespace that cannot affect functionality (errors should be reserved for functional flaws) but should never be committed.
- Adding `--ext .js,.json,.html` and adding eslint-plugin-html (eslint-plugin-json is already present) to lint our JSON and script tags in HTML files.
- Replace `'src/**/*.js'` with `.` and use an .eslintignore ([example](https://github.com/wikimedia/wikimedia-page-library/blob/master/.eslintignore)) so we by default lint everything we can.

joakin replied:
Great advice, thanks! I'm submitting the changes as new commits on the same PR to not deal with rebasing and force pushing losing context for the comments.

* ad7e5b2 - changing the lint script to adopt the lint:all advice
* eb87366 - add --cache to the lint script & add .eslintcache to gitignore
* 8128fd1 - add --max-warnings 0 to lint script

Re: globs, Most programs understand string globs based on the package [node-glob](https://github.com/isaacs/node-glob) so it is better to pass a string than to rely on the shell's globbing (for example npm uses sh to run the scripts, which doesn't have proper globstar support).

Regarding `--ext .js,.json,.html` it seems overkill to set up linting for them given we don't have any files of that kind, feel free to submit a followup commit, I'd be happy to merge.

Re: Replace 'src/**/*.js', I believe that having the explicit path to the JS files there kind of makes it easy to check the paths of the files for other tools like prettier. Happy to re-evaluate later if we increase the amount of eslint config

niedzielski replied:
Thanks for the node-glob tip!

> Regarding --ext .js,.json,.html it seems overkill to set up linting for them given we don't have any files of that kind, feel free to submit a followup commit, I'd be happy to merge.
We do have package.json but I suppose we won't be able to run any scripts if that's broken.

There's some other feedback on eslint-plugin-json so let's punt!

> Re: Replace 'src/**/*.js', I believe that having the explicit path to the JS files there kind of makes it easy to check the paths of the files for other tools like prettier. Happy to re-evaluate later if we increase the amount of eslint config

I don't mind punting on this too. My thoughts were that I prefer linting to default to complaining rather abstaining. I might try to make a more compelling argument in a future PR.

niedzielski commented on package.json:
```diff
@@ -0,0 +1,57 @@
+  "version": "0.0.0",
+  "description": "An API driven skin for MediaWiki",
+  "scripts": {
+    "start": "run-p -n dev:*",
```
For my own education, I'm going to take this opportunity to ask: is it good practice to use `-s` in all NPM scripts that call other NPM scripts? For example, `npm run -s` and `run-p -s`. NPM has kind of verbose output and I've been struggling with trying to understand whether it's better to put this setting in my personal `~/.npmrc` or each script I add.

joakin replied:
I personally set it with `npm config set loglevel silent` and change it as needed to higher levels, but I don't have a strong opinion on this.

It could be worth noting also that maybe even if you set -s here, if you don't have it when you do `npm start` you will probably get the obnoxious message if the error code is >0 (I haven't checked tho).

I say let's leave it like this, if it becomes annoying I'm happy to review or submit a patch to change it. Would that be ok?

niedzielski replied:
Thanks, that sounds good. I followed suit in my patches.

niedzielski replied:
(Oh, and I was only asking because I wanted your advisement! Thanks for the detail!)

I just noticed, and we can hit this in a future patch for sure, `dev:*` should be escaped to avoid possible shell completion.

niedzielski commented on src/server/index.js:
```diff
@@ -0,0 +1,20 @@
+const express = require("express");
```
Any reason not to use `import`s or will these come when we switch to Babel?

joakin replied:
When we switch to babel we will have to decide if we want to transpile these node.js only files.

I'd lean towards transpiling as little as we need, but I'm open to anything. Let's discuss in person or when we reach that point.

niedzielski commented on .eslintrc.json:
```diff
@@ -0,0 +1,20 @@
+    "prettier"
+  ],
+  "plugins": [
+    "prettier"
```
eslint-plugin-json is specified as dependency but must be specified as a plugin here.

joakin replied:
eslint-plugin-json was brought up because eslint-config-node-services or the wikimedia one had it as a peer dependency. If you think it is worth using, I'd appreciate if you can send a commit configuring it, if the node-services preset or the wikimedia one are not actually configuring it. I've never used it before.

niedzielski replied:
I've found it useful as JSON files often exist side-by-side with JavaScript. I'll try to make a PR for this and see if I can convince you.

niedzielski commented on .eslintrc.json:
```diff
@@ -0,0 +1,20 @@
+  },
+  "env": {
+    "es6": true,
+    "node": true
```
Do we want to enable the Node environment for the whole project or should we add a sub-.eslintrc.json that enables node for one directory? I'm asking because it seems in some areas we may want a browser environment and others a Node environment?

joakin replied:
Most definitely. I haven't dealt with that before so I'd like to discuss some options about how to do it if you don't mind.

I've created https://phabricator.wikimedia.org/T172683 to discuss and not forget

niedzielski commented on package.json:
```diff
@@ -0,0 +1,57 @@
+      "npm run test"
+    ]
+  },
+  "repository": {
```
Consider also `"repository": "github:niedzielski/marvin"`.

joakin replied:
Cool! 👌 I didn't know about that! I just read https://docs.npmjs.com/files/package.json#repository and it seems it is only for the shortcuts. Given we are going to have to move repository to gerrit (☹️), I say we don't touch it for now, and just change the URL when we move away

niedzielski commented on readme.md:
```diff
@@ -1,4 +1,13 @@
\ No newline at end of file
+## Development
+
+Marvin requires the Node.js active LTS version, and npm > 5.X.
```
How about we just add a `"engines": { "node": "^5.0.0" }` to package.json and remove this line?

joakin replied:
Great idea. Addressed in 29f3824, also the npm version.

I've kept some info in the README because it is a bit more friendly than just linking people away from the documentation into a README file.

niedzielski commented on src/server/templates/page.js:
```diff
@@ -0,0 +1,16 @@
+module.exports = function page({ body = "", assets, title }) {
```
Any reason not to favor arrow syntax here (you wouldn't even need return or brackets)? Since this is our very first function that we've defined, would you mind adding a JSDoc with the typing? This will also give us a chance to test the JSDoc linting plugin you've added! Something like:

```js
/**
 * @arg {?string} body HTML
 * @arg {!{entry: ?string, asset: ?object}} assets Manifest of filename entry points to bundled assets.
 * @arg {?string} title
 * @return {!string} HTML
 */```

joakin replied:
I generally prefer to use named functions on the top level variables of the scope, to have some names to refer to in case of having to debug, etc.

If you have strong opinions I don't mind changing it.

---

Re docs, sorry I totally forgot, I was on bootstrapping mode. Addressed here e867bd0. Saw the linting fail as I was adding the docs 👌

niedzielski commented on src/server/templates/page.js:
```diff
@@ -0,0 +1,16 @@
+module.exports = function page({ body = "", assets, title }) {
+  return `
```
I know this file is just to get the ball rolling(!) but I'd be curious to know if this is the plan for how we'll be writing all HTML templates or if we should investigate a Hogan / Handlebars alternative.

joakin replied:
For the moment it is just to get going. You are right, if we need to have more server side templates then we should definitely swap to Handlebars. 👍

I'm assuming the whole page template will be the only one we will need. The rest of the rendering of views should be happening withing preact or whatever.

niedzielski commented on package-lock.json:
```diff
@@ -0,0 +1,5493 @@
+{
```
Yayyyy!!

niedzielski commented on package-lock.json:
```diff
@@ -0,0 +1,5493 @@
+      "requires": {
+        "anymatch": "1.3.2",
+        "async-each": "1.0.1",
+        "fsevents": "1.1.2",
```
Would you mind double checking this? When I install locally, fsevents is omitted.

joakin replied:
Let's troubleshoot in person, this file is autogenerated by npm 😕 . Do you have the latest npm@5? I'm on npm 5.3.0

niedzielski replied:
I'm on v5.3.0 as well (and I still see this issue on the latest PR) :[

niedzielski commented on package.json:
```diff
@@ -0,0 +1,57 @@
+  },
+  "lint-staged": {
+    "*.js": [
+      "npm run test"
```
Consider the shorter `npm test` too.

joakin replied:
Addressed in 2d63c10

joakin committed:
Add script lint:all

And keep script lint as the base eslint command.

That will allow us to call linting with specific file paths, like:

npm run lint -- --fix src/server/blah.js

For running all files you do now npm run lint:all

joakin committed:
Add --cache to eslint to only lint changed files

joakin committed:
Add --max-warnings 0 to eslint script

That way when linting even warnings will trigger a non-zero exit code.

joakin committed:
Clean up .gitignore

As per review comments in:
https://github.com/niedzielski/marvin/pull/1#discussion_r131392283

Remove unused boilerplate from .gitignore and only keep exactly what we
need.

joakin committed:
Move md link to its section in README

joakin committed:
Add documentation info and link to CHANGELOG

joakin committed:
Specify node and npm versions in package.json

Also update README to refer to package.json#engines

joakin committed:
Document server/templates/page.js

joakin committed:
Hygiene: Use npm test instead of npm run test

joakin commented:
@niedzielski  Sent up some extra commits addressing the points in the review. Added links to them in the review comments too.

niedzielski commented on src/server/templates/page.js:
```diff
@@ -0,0 +1,24 @@
+ * @param {?string} params.title Title of the page
+ * @param {string} [params.body = ""] HTML to render in the body of the page
+ * @param {Object.<string, Object.<string, string>>} params.assets Manifest of filename entry points to bundled assets.
+ * @return {string} HTML
```
I find that marking the nullity expectations to greatly enhance documentation. Let's talk about we want to annotate future JSDocs!

joakin replied:
Oh damn I didn't know about that, I wrongly assumed that no marker was the same as the ! marker. I've added it to my TODO list to fix this jsdoc

joakin replied:
Thanks!

niedzielski commented on package.json:
```diff
@@ -0,0 +1,54 @@
+    "type": "git",
+    "url": "git+https://github.com/niedzielski/marvin.git"
+  },
+  "keywords": ["mediawiki", "wikimedia", "web", "node", "pwa"],
```
I love this terse style! However, please note that `npm install` will undo it!

joakin replied:
Argh, goddammit npm. This was written by `npm init` after filling up the cli stuff...  Needs fixing, I'll add it to TODO.

niedzielski commented on .gitignore:
```diff
@@ -0,0 +1,7 @@
+# Compiled assets
```
I recommend anchoring these paths to root for extra precision but we can do that another day.

joakin replied:
Never done it, not against it 😛

niedzielski commented:
Looks great!! Thanks for bearing with me!
